### ISSUE TITLE
Only lints and shows errors for most recently saved file in watch mode

### ISSUE THREAD

**Issue by densk1** [webpack-contrib/eslint-webpack-plugin#44]
- Operating System: OSX 10.15.4
- Node Version: v12.18.0
- NPM Version: 6.14.8
- webpack Version: 5.4.0
- eslint-webpack-plugin Version: 2.3.0

### Expected Behavior

When in watch mode, each time I save a file, all files in the tree are linted.

### Actual Behavior

When in Webpack watch mode, on start it lints and shows errors for both `A.ts` and file `B.ts` in the linting path. When I save `A.ts` it only checks and displays errors for `A.ts`. When I save `B.ts`, , it then only lints and shows errors for file B.

### Code

```js
// webpack.config.js
...
  plugins: [
    new CleanWebpackPlugin(),
    new ESLintPlugin({
      files: 'src/**/*',
      // formatter: 'codeframe', // table
      extensions: ['js', 'ts', 'tsx'],
      // cache: true,
      failOnError: false,
      failOnWarning: false,
      emitWarning: true,
    }),
  ],
...
```

### How Do We Reproduce?

Switch between files with linting errors in them, and save each one while in watch mode on `v2.3.0`.

### Fix / Work around

I have reverted back to `v2.1.0`

**Comment by densk1** [comment]
Guessing, by reading changes, but is it possibly related to this:

https://github.com/webpack-contrib/eslint-webpack-plugin/blob/8287872f07853590921309ebca2b3cd05f4a73d3/src/index.js#L28-L38

and this

https://github.com/webpack-contrib/eslint-webpack-plugin/blob/8287872f07853590921309ebca2b3cd05f4a73d3/src/index.js#L43-L50

Blocking re-running of `this.run`?

**Comment by jsg2021** [comment]
There is a watch mode test. Could you make a test case that fails?

**Comment by jsg2021** [comment]
> Guessing, by reading changes, but is it possibly related to this:
> 
> 
> 
> https://github.com/webpack-contrib/eslint-webpack-plugin/blob/8287872f07853590921309ebca2b3cd05f4a73d3/src/index.js#L28-L38
> 
> 
> 
> and this
> 
> 
> 
> https://github.com/webpack-contrib/eslint-webpack-plugin/blob/8287872f07853590921309ebca2b3cd05f4a73d3/src/index.js#L43-L50
> 
> 
> 
> Blocking re-running of `this.run` for watch?

The run method is probably miss named. It's actually registering hooks to a compiler. For watch mode, that compiler lives for the duration of your watch. So those hooks are already present it skips and let's the original hooks process each compilation.

**Comment by densk1** [comment]
> There is a watch mode test. Could you make a test case that fails?

I've modified `watch.test.js`. `expect(message).toEqual(expect.stringMatching('prefer-const'));` fails in `thirdPass()`. It passes in `secondPass()` but the file has not changed.

```
// watch.test.js

import { join } from 'path';
import { writeFileSync } from 'fs';

import { removeSync } from 'fs-extra';

import pack from './utils/pack';

const target = join(__dirname, 'fixtures', 'watch-entry.js');
const target2 = join(__dirname, 'fixtures', 'watch-entry-2.js');
const targetExpectedPattern = expect.stringMatching(
  target.replace(/\\/g, '\\\\')
);

describe('watch', () => {
  afterEach(() => {
    removeSync(target);
  });

  it('should watch', (done) => {
    const compiler = pack('good');

    const watch = compiler.watch({}, (err, stats) => {
      watch.close();
      expect(err).toBeNull();
      expect(stats.hasWarnings()).toBe(false);
      expect(stats.hasErrors()).toBe(false);
      done();
    });
  });

  it('should watch with unique messages', (done) => {
    writeFileSync(target, 'var foo = stuff\n');

    let next = firstPass;
    const compiler = pack('watch');
    const watch = compiler.watch({}, (err, stats) => next(err, stats));

    function firstPass(err, stats) {
      expect(err).toBeNull();
      expect(stats.hasWarnings()).toBe(false);
      expect(stats.hasErrors()).toBe(true);
      const { errors } = stats.compilation;
      expect(errors.length).toBe(1);
      const [{ message }] = errors;
      expect(message).toEqual(targetExpectedPattern);
      expect(message).toEqual(expect.stringMatching('\\(3 errors,'));

      next = secondPass;

      writeFileSync(target2, "let bar = false;\n");
      writeFileSync(target, "const x = require('./watch-entry-2.js')\n\nconst foo = false;\n");
    }

    function secondPass(err, stats) {
      expect(err).toBeNull();
      expect(stats.hasWarnings()).toBe(false);
      expect(stats.hasErrors()).toBe(true);
      const { errors } = stats.compilation;
      expect(errors.length).toBe(1);
      const [{ message }] = errors;
      expect(message).toEqual(targetExpectedPattern);
      expect(message).toEqual(expect.stringMatching('no-unused-vars'));
      expect(message).toEqual(expect.stringMatching('prefer-const')); // `prefer-const` passes here
      expect(message).toEqual(expect.stringMatching('\\(4 errors,'));
      
      next = thirdPass;
      
      writeFileSync(target, "const x = require('./watch-entry-2.js')\nconst foo = 0\n");
    }
    
    function thirdPass(err, stats) {
      expect(err).toBeNull();
      expect(stats.hasWarnings()).toBe(false);
      expect(stats.hasErrors()).toBe(true);
      const { errors } = stats.compilation;
      expect(errors.length).toBe(1);
      const [{ message }] = errors;
      expect(message).toEqual(targetExpectedPattern);
      expect(message).toEqual(expect.stringMatching('no-unused-vars'));
      expect(message).toEqual(expect.stringMatching('prefer-const')); // `prefer-const` fails here
      expect(message).toEqual(expect.stringMatching('\\(1 error,'));

      next = finish;

      writeFileSync(
        target,
        '/* eslint-disable no-unused-vars */\nconst foo = false;\n'
      );
    }

    function finish(err, stats) {
      watch.close();
      expect(err).toBeNull();
      expect(stats.hasWarnings()).toBe(false);
      expect(stats.hasErrors()).toBe(false);
      done();
    }
  });
});

```

**Comment by refaelok** [comment]
I have the exact same issue.
I'm following this issue and waiting for the next fix.

**Comment by inker** [comment]
Still occurring in 2.4.0.

**Comment by ricardogobbosouza** [comment]
@inker 
I haven't released 2.4.1 yet :grimacing:

**Comment by jsg2021** [comment]
@ricardogobbosouza I fixed one last edge case for this

**Comment by refaelok** [comment]
actually the issue was fixed but created another performance issue.
now for each change in some file the build open new nodejs process and take more memory from cpu, that mean while you cding and saving in big projects, you get performance issue, due to for 3 with interval of few seconds create performance issue.

**Comment by jsg2021** [comment]
the thread pool is now disabled by default. As a bonus, the pool will only be used for the initial completion.

**Comment by inker** [comment]
Works in 2.4.1. Thanks.
